osquery/filesystem/linux/proc.cpp

```diff
@@ -16,54 +16,165 @@
+
+    char link_destination[PATH_MAX] = {};
+    auto link_dest_length =
+        readlink(symlink_path.data(), link_destination, PATH_MAX - 1);
+    if (link_dest_length < 0) {
+      return Status(1,
+                    std::string("Failed to retrieve the inode for namespace ") +
+                        namespace_name + " in process " + process_id);
+    }
+
+    // The link destination must be in the following form: namespace:[inode]
+    if (std::strncmp(link_destination,
+                     namespace_name.data(),
+                     namespace_name.size()) != 0 ||
+        std::strncmp(link_destination + namespace_name.size(), ":[", 2) != 0) {
+      return Status(1,
+                    std::string("Invalid descriptor for namespace ") +
+                        namespace_name + " in process " + process_id);
     }
-  } catch (const boost::filesystem::filesystem_error& e) {
-    VLOG(1) << "Exception iterating Linux processes " << e.what();
-    return Status(1, e.what());
+
+    // Parse the inode part of the string; strtoull should return us a pointer
+    // to the
+    // closing square bracket
+    const char* inode_string_ptr = link_destination + namespace_name.size() + 2;
+    char* square_bracket_ptr = nullptr;
+
+    auto inode = static_cast<ino_t>(
+        std::strtoull(inode_string_ptr, &square_bracket_ptr, 10));
+    if (inode == 0 || square_bracket_ptr == nullptr ||
+        *square_bracket_ptr != ']') {
+      return Status(
+          1,
+          std::string("Invalid inode value in descriptor for namespace ") +
+              namespace_name + " in process " + process_id);
+    }
+
+    namespace_list[namespace_name] = inode;
   }
 
   return Status(0, "OK");
 }
 
+std::string procDecodeAddressFromHex(const std::string& encoded_address,
+                                     int family) {
+  char addr_buffer[INET6_ADDRSTRLEN] = {0};
+  if (family == AF_INET) {
+    struct in_addr decoded;
+    if (encoded_address.length() == 8) {
+      sscanf(encoded_address.c_str(), "%X", &(decoded.s_addr));
+      inet_ntop(AF_INET, &decoded, addr_buffer, INET_ADDRSTRLEN);
+    }
+  } else if (family == AF_INET6) {
+    struct in6_addr decoded;
+    if (encoded_address.length() == 32) {
+      sscanf(encoded_address.c_str(),
+             "%8x%8x%8x%8x",
+             (unsigned int*)&(decoded.s6_addr[0]),
+             (unsigned int*)&(decoded.s6_addr[4]),
+             (unsigned int*)&(decoded.s6_addr[8]),
+             (unsigned int*)&(decoded.s6_addr[12]));
+      inet_ntop(AF_INET6, &decoded, addr_buffer, INET6_ADDRSTRLEN);
+    }
+  }
+
+  return std::string(addr_buffer);
+}
+
+unsigned short procDecodePortFromHex(const std::string& encoded_port) {
+  unsigned short decoded = 0;
+  if (encoded_port.length() == 4) {
+    sscanf(encoded_port.c_str(), "%hX", &decoded);
+  }
+  return decoded;
+}
+
+Status procProcesses(std::set<std::string>& processes) {
+  auto L_procProcessesCallback = [](const std::string& process_id,
```

Comment:
Where does this `L_NAME` for lambdas come from, I've never seen this naming scheme.